### PR TITLE
Add prestart-hook for device plugin

### LINKED ISSUE #383: Feature: Deleting memory of prevous GPU runs before running the existing job
### 1. Issue or feature description
In my company we found that it is possible for users to view the memory of previous jobs.  We forked this repo and added a prehook that will delete existing memory before running a pod.  

Is there interest in opening this up as an enhancement to this repo?  We added it as a separate cmd as part of the device-plugin but we are open to designing this to be a more integration solution.

### PR DESCRIPTION
**Pull request by kannon92**
Fixes #383 

We wanted to see what NVIDIA thinks of this addition to the k8s-device-plugin.  

We wanted to provide a way to delete the memory of a previous node and this PR achieves that.  We use this in our production cluster and we wanted to get it merged upstream.  

We provided a way to inject custom prehook commands so a separate caller can do a purge of memory if needed.

Happy to leave this as a discussion item!

### CODE REVIEW

**Comment by elezar** [comment]
Hi @kannon92. Thanks for the PR. Note that this is currently a read-only mirror of https://gitlab.com/nvidia/kubernetes/device-plugin and as such an MR should be created there instead.

File: cmd/nvidia-device-plugin/prestart-hook.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+		for _, device := range devices {
+			preStartInfo, err := hook.getDevicePreStartInfo(ctx, device.ID, logCtx)
+			if err != nil {
+				log.Printf("PreStartHook(%s): %v", logCtx, err)
+				continue
+			}
+
+			failures := preStartInfo.incPreStartFailures()
+			log.Printf("PreStartHook(%s): Failure count for device %s: %d\n", logCtx, device.ID, failures)
+			if failures > 4 {
```
Comment (elezar) [inline cmd/nvidia-device-plugin/prestart-hook.go:98]: Why is `4` the threshold? Should this not be dependent on the hook?

File: cmd/nvidia-device-plugin/prestart-hook.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		log.Printf("PreStartHook(%s): Creating device PreStart info for %q\n", logCtx, deviceId)
+		info = newDevicePreStartInfo()
+		hook.deviceInfos[deviceId] = info
+	}
+	return info, nil
+}
+
+func findDevice(deviceId string, allDevices []*rm.Device) (*rm.Device, error) {
+	for _, device := range allDevices {
+		if device.ID == deviceId {
```
Comment (elezar) [inline cmd/nvidia-device-plugin/prestart-hook.go:238]: Not that with GPU sharing enabled, `device.ID` may not be the `uuid` of the device. There is ` GetUUID` method that can be used instead.

File: cmd/nvidia-device-plugin/server.go [hunk 3]
```diff
@@ -343,8 +358,14 @@ func (plugin *NvidiaDevicePlugin) getAllocateResponseForCDIAnnotations(responseI
 	return &response
 }
 
-// PreStartContainer is unimplemented for this plugin
-func (plugin *NvidiaDevicePlugin) PreStartContainer(context.Context, *pluginapi.PreStartContainerRequest) (*pluginapi.PreStartContainerResponse, error) {
+func (plugin *NvidiaDevicePlugin) PreStartContainer(ctx context.Context, r *pluginapi.PreStartContainerRequest) (*pluginapi.PreStartContainerResponse, error) {
+	if plugin.preStartRequired() {
+		timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
```
Comment (elezar) [inline cmd/nvidia-device-plugin/server.go:363]: Should the timeout be configurable?

File: cmd/nvidia-device-plugin/prestart-hook.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+func (hook *PreStartHook) runPreStartContainerCommand(ctx context.Context, devices []*rm.Device) (unhealthy []string, err error) {
+	if len(devices) == 0 {
+		return
+	}
+	logCtx := getLogContextFromDevices(devices)
+
+	cmd := hook.preStartCmd[0]
+	args := hook.preStartCmd[1:]
+	for _, device := range devices {
+		uuid := strings.TrimPrefix(device.ID, "GPU-")
```
Comment (elezar) [inline cmd/nvidia-device-plugin/prestart-hook.go:139]: Does this mean that we don't support MIG devices?

This also adds a specific API for the prestart commands which is not documented.

**Review by elezar — COMMENTED** [review]
@kannon92 as mentioned, please submit an MR against GitLab instead since this is where we review contributions (from ourselves an the community).

One high-level comment that I will make here is whether this is the right type of extension to the device plugin. We are in the process of rolling out support for generating CDI specifications for the available devices in a system and using CDI as the mechanism for requesting devices instead of our existing mechanisms.

Since the CDI specification allows for per-Device (resource) hooks to be defined, it seems as if this would be a more stable way to ensure that, for example, memory is cleared. This would mean that the CDI specification that is generated by the device plugin is augmented to include additional hooks, but may lead to a more flexible solution.

With that said, if there is a motivation for using the device plugin API-based approach here, we can continue discussion on the MR and determine whether this is something that can be accepted as an extension.

**Comment by elezar** [comment]
> > Hi @kannon92. Thanks for the PR. Note that this is currently a read-only mirror of https://gitlab.com/nvidia/kubernetes/device-plugin and as such an MR should be created there instead.
> 
> Oh okay. Mind If I open a PR on this repo that states that? I had no idea!

Sure. We had a bot that would respond to PRs, but a note in the readme would be a good start.

**Comment by kannon92** [comment]
> > > Hi @kannon92. Thanks for the PR. Note that this is currently a read-only mirror of https://gitlab.com/nvidia/kubernetes/device-plugin and as such an MR should be created there instead.
> > 
> > 
> > Oh okay. Mind If I open a PR on this repo that states that? I had no idea!
> 
> Sure. We had a bot that would respond to PRs, but a note in the readme would be a good start.

https://github.com/NVIDIA/k8s-device-plugin/pull/393

**Comment by kannon92** [comment]
> @kannon92 as mentioned, please submit an MR against GitLab instead since this is where we review contributions (from ourselves an the community).
> 
> One high-level comment that I will make here is whether this is the right type of extension to the device plugin. We are in the process of rolling out support for generating CDI specifications for the available devices in a system and using CDI as the mechanism for requesting devices instead of our existing mechanisms.
> 
> Since the CDI specification allows for per-Device (resource) hooks to be defined, it seems as if this would be a more stable way to ensure that, for example, memory is cleared. This would mean that the CDI specification that is generated by the device plugin is augmented to include additional hooks, but may lead to a more flexible solution.
> 
> With that said, if there is a motivation for using the device plugin API-based approach here, we can continue discussion on the MR and determine whether this is something that can be accepted as an extension.

I'm really interested in this!  Is this related to the DRA work going in Kubernetes?

**Comment by kannon92** [comment]
@elezar I opened https://gitlab.com/nvidia/kubernetes/device-plugin/-/merge_requests/278 so should we close this one?

I tried to address some of your comments in this PR but not all of them.

**Comment by kannon92** [comment]
Closing this because of gitlab PR

**Pull request closed without merging**